### PR TITLE
Templating: Migrate Template variables to Redux

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
This PR creates a first attempt for a pattern to be used in migrating existing template variables to Redux whilst keeping all the Angular directives and *Srv working as before. Using `getters` and `setters` to hide the fact values are stored in Redux.
The thinking is that when this is done for all Variable types we can start focusing on removing `variableSrv` and replacing Angular components in PRs to come with React components.

**Which issue(s) this PR fixes**:
Part of: #19896

**Special notes for your reviewer**:
~~I've only implemented the pattern for QueryVariables. No other Variable will work with this PR.~~
Pattern is now implemented for all Variable types.

My plans are to remove the `temporary` and `isTemporary` complexity in future PRs when the Editor has been migrated to React and using the local state of the that React component instead.

### CODE REVIEW

File: public/app/features/templating/editor_ctrl.ts [hunk 1]
```diff
@@ -174,7 +175,7 @@ export class VariableEditorCtrl {
 
     $scope.reset = () => {
       $scope.currentIsNew = true;
-      $scope.current = variableSrv.createVariableFromModel({ type: 'query' });
+      $scope.current = variableSrv.createVariableFromModel({ type: 'query', isTemporary: true });
```
Comment (hugohaggmark) [inline public/app/features/templating/editor_ctrl.ts]: `isTemporary` tells the constructor for this VariableType that this isn't to be stored in ReduxState. Variables are stored in ReduxState when load a dashboard and when the user clicks add in the Editor.

File: public/app/features/templating/query_variable.ts [hunk 2]
```diff
@@ -76,13 +62,141 @@ export class QueryVariable implements QueryVariableModel, VariableActions {
     private timeSrv: TimeSrv
   ) {
     // copy model properties to this instance
-    assignModelProperties(this, model, this.defaults);
+    if (model.isTemporary) {
+      this.temporary = {} as QueryVariableModel;
```
Comment (hugohaggmark) [inline public/app/features/templating/query_variable.ts]: This is a `temporary` model => this is when user clicks new in the Editor

**Comment by ryantxu** [comment]
@hugohaggmark -- does #16128 relate?

**Comment by hugohaggmark** [comment]
> @hugohaggmark -- does #16128 relate?

Sort of relates but it doesn't feel like a great solution to me.

File: public/app/features/templating/query_variable.ts (hunk 2 above)
Comment (torkelo) [reply public/app/features/templating/query_variable.ts]: if we implemented the solution to this UX issue we can get rid of the temporary handling: https://github.com/grafana/grafana/issues/20703  , worth doing that before merging this?

Comment (hugohaggmark) [reply public/app/features/templating/query_variable.ts]: My next PR will migrate the Editor to React so it will be removed then at least in theory

Comment (hugohaggmark) [reply public/app/features/templating/query_variable.ts]: Maybe after the next PR is done we could get a clearer picture if this is a valid way of migrating the template variables to Redux?

File: public/app/features/templating/TextBoxVariable.ts [hunk 3]
```diff
@@ -29,14 +27,72 @@ export class TextBoxVariable implements TextBoxVariableModel, VariableActions {
...
   /** @ngInject */
   constructor(private model: any, private variableSrv: VariableSrv) {
-    assignModelProperties(this, model, this.defaults);
+    if (model.useTemporary) {
+      this.temporary = {} as QueryVariableModel;
+      assignModelProperties(this.temporary, model, this.defaults);
+      this.id = -1;
+    } else {
+      this.temporary = null;
+      this.id = createVariableInState(model, this.defaults);
```
Comment (torkelo) [inline public/app/features/templating/TextBoxVariable.ts]: this part feels a bit wrong, that the constructor is having a redux state side-effect

**Review by torkelo — APPROVED** [review]
Been looking at this a few times, think it's a good approach that allows us to have an up to date model in redux, some funky stuff but it's temporary. 

But we can merge after 6.6 and maybe once we have another PR that this depends on

**Comment by mckn** [comment]
Have looked at this PR and it looked good. I think I understand what we want to achieve but I think we might be able to change this around by slicing this feature differently.

I don't have the whole picture but would love to have a chat and see if my input can help slicing it differently @hugohaggmark @torkelo.

**Comment by hugohaggmark** [comment]
Closing this PR in favor of #21873

**Pull request closed without merging**